src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -339,11 +344,32 @@ private ImmutableArray<SyntaxTree> ParseAdditionalSources(ISourceGenerator gener
             var prefix = GetFilePathPrefixForGenerator(generator);
             foreach (var source in generatedSources)
             {
-                trees.Add(ParseGeneratedSourceText(source, Path.Combine(prefix, source.HintName), cancellationToken));
+                trees.Add(GetOrParseGeneratedSourceText(source, Path.Combine(prefix, source.HintName), cancellationToken));
             }
             return trees.ToImmutableAndFree();
         }
 
+        private SyntaxTree GetOrParseGeneratedSourceText(GeneratedSourceText input, string fileName, CancellationToken cancellationToken)
+        {
+            if (s_parsedGeneratedSources.TryGetValue(input.Text, out var existingTree)
+                && Equals(_state.ParseOptions, existingTree.Options)
+                && Equals(fileName, existingTree.FilePath))
```

Comment:
Please use an explicit comparer when comparing `string` values.

Reply:
➡️ This is now fixed.